### ISSUE TITLE
should Chain properties depends on fantasy-land/ap (and possibly .map)?

### ISSUE THREAD

**Issue by bunglegrind** [fantasyland/fantasy-land#334]
It's more like a question than an actual issue:

Since Chain must implement also the Apply specification, I'm expecting that one of the Chain properties should depends on something related to the Apply properties (as, for instance, Applicative's `fantasy-land/of` properties resort to the `fantasy-land/ap` method).

**Comment by Avaq** [comment]
In my personal ([contested](https://github.com/gcanti/fp-ts/issues/1173)) view it does, but not as explicitly. The dependency is captured in the [Derivations section of the spec](https://github.com/fantasyland/fantasy-land#derivations), where a relationship between `ap` and `chain` is described:

> [fantasy-land/ap](https://github.com/fantasyland/fantasy-land#ap-method) may be derived from [fantasy-land/chain](https://github.com/fantasyland/fantasy-land#chain-method):
>
>     function ap(m) { return m['fantasy-land/chain'](f => this['fantasy-land/map'](f)); }

I read this as: "A Monad must implement chain such that `ap` may be derived from it, and therefore every Monad has `ap`, and `ap` must behave as if derived from `chain`".

**Comment by bunglegrind** [comment]
Well, the derivation section uses the "may" expression, not the "must". IMHO, an explicit property should be added to the `fantasy-land/chain` method, which should enfasize the relationship between `fantasy-land/ap` and `fantasy-land/chain` (and, possibly, `fantasy-land/map`). Otherwise, Chain should be placed at the same level of Functor

**Comment by Avaq** [comment]
I agree with you. I think the "may"-expression in the derivations section is a source of confusion. It's using "may" to suggest that a developer doesn't _have_ to derive their implementation of `ap` from their implementation of `chain`, but "may" choose do so.

The reason they "may" choose to do so, is because it is assumed (but never explicitly stated) that this is always possible. I think the spec should be updated to make this more clear.

Furthermore, it is never made clear whether the non-derived implementation of `ap` must behave in exactly the same way as the version derived from `chain` would have, but in my view it should. This is the view that's contested (as far as I know, also in the broader FP community).

-----

I've seen this discussion play out in other threads on this repo as well. A quick search came up with https://github.com/fantasyland/fantasy-land/issues/214 but I think there might be more.

**Comment by bunglegrind** [comment]
I see. If I look through all the definitions I find other examples (traversable, for instance). Anyway, I'll follow your suggestion: I'll consider the derivation section mandatory. Thanks!
